core/pkg/artifacts/saver.go

```diff
@@ -18,18 +19,102 @@ import (
 	"github.com/wandb/wandb/core/internal/filetransfer"
 	"github.com/wandb/wandb/core/internal/gql"
 	"github.com/wandb/wandb/core/internal/hashencode"
+	"github.com/wandb/wandb/core/internal/namedgoroutines"
 	"github.com/wandb/wandb/core/internal/nullify"
 	"github.com/wandb/wandb/core/internal/observability"
 	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
 )
 
+// uploadBufferPerArtifactName is the number of Save operations
+// that can be queued per artifact name before Save begins to block.
+//
+// The value is high enough so that Save almost never blocks,
+// without consuming too much memory just for bookkeeping.
+const uploadBufferPerArtifactName = 32
+
+// ArtifactSaver manages artifact uploads.
 type ArtifactSaver struct {
+	logger              *observability.CoreLogger
+	graphqlClient       graphql.Client
+	fileTransferManager filetransfer.FileTransferManager
+	fileCache           Cache
+
+	// uploadsByName ensures that uploads for the same artifact name happen
+	// serially, so that version numbers are assigned deterministically.
+	uploadsByName *namedgoroutines.Operation[*ArtifactSaveContext]
+}
+
+func NewArtifactSaver(
+	logger *observability.CoreLogger,
+	graphqlClient graphql.Client,
+	fileTransferManager filetransfer.FileTransferManager,
+) *ArtifactSaver {
+	return &ArtifactSaver{
+		logger:              logger,
+		graphqlClient:       graphqlClient,
+		fileTransferManager: fileTransferManager,
+		fileCache:           NewFileCache(UserCacheDir()),
+		uploadsByName: namedgoroutines.New(
+			uploadBufferPerArtifactName,
+			func(saveCtx *ArtifactSaveContext) {
+				artifactID, err := saveCtx.Save()
+				saveCtx.resultChan <- ArtifactSaveResult{
+					ArtifactID: artifactID,
+					Err:        err,
+				}
+				close(saveCtx.resultChan)
+			},
+		),
+	}
+}
+
+type ArtifactSaveResult struct {
+	ArtifactID string
+	Err        error
+}
+
+// Save asynchronously uploads an artifact.
+//
+// This returns a channel to which the operation result is eventually
+// pushed. The channel is guaranteed to receive exactly one value and
+// then close.
+func (as *ArtifactSaver) Save(
+	ctx context.Context,
+	artifact *spb.ArtifactRecord,
+	historyStep int64,
+	stagingDir string,
+) <-chan ArtifactSaveResult {
+	resultChan := make(chan ArtifactSaveResult, 1)
+
+	as.uploadsByName.Go(
+		artifact.Name,
+		&ArtifactSaveContext{
+			ctx:                 ctx,
+			logger:              as.logger,
+			graphqlClient:       as.graphqlClient,
+			fileTransferManager: as.fileTransferManager,
+			fileCache:           as.fileCache,
+			artifact:            artifact,
+			historyStep:         historyStep,
+			stagingDir:          stagingDir,
+			maxActiveBatches:    5,
+			resultChan:          resultChan,
+		},
+	)
+
+	return resultChan
+}
+
+// ArtifactSaveContext is a save operation for one artifact.
+type ArtifactSaveContext struct {
```

Comment:
Calling this a 'context' is a little confusing to me because I usually think of contexts as long-lived objects that store things like loggers and connections, and inside contexts something else does the work. Here the atom of work is the context. No blocking comment here, go with whatever makes most sense to you, but I would have the ArtifactSaver named ArtifactSaveManager or something and ArtifactSaveContext as ArtifactSaver (as it was originally).

Reply:
Good point, I prefer your naming. I updated the PR!

I avoid the word "Manager" in type names because it often doesn't mean much and is easy to overuse, but in this case I couldn't come up with a better name for what I formerly called `ArtifactSaveContext`. Maybe `ArtifactSaveOperation`, but it's not as clear as the pair `ArtifactSaveManager`/`ArtifactSaver`.

Reply:
Agreed. As the aphorism goes, naming is one of the only two hard things in computer science. Every term is so overused as to become meaningless, but there so often aren't better ones either.